### DISCUSSION TITLE
请问immortalwrt和lean的lede、官方的openwrt有什么不同点？

### DISCUSSION THREAD

**Discussion by WROIATE** [immortalwrt/immortalwrt#1109] [Q&A] [answered]
一直以来的疑惑，没能有时间去好好研究代码，看到目前各种维护分支，难免有些疑惑他们之间的区别。希望有知道细节的开发人员能解答下

**Comment by BabXhwZUNBeTV** [comment]


**Reply by WROIATE** [reply]
@BabXhwZUNBeTV 
朋友，我不清楚我的这个话题为什么触怒了你，我在这里先道歉

然后我解释下，“没时间”是指相对意义上的，就像我在这里发帖，或者看readme，这些时间和看代码的时间相比较是比较短的，事实上我也确实能接受并且这么做了；但看代码可不是，我不仅需要相关c代码、编译的专业知识，还要对除官方源码外的patch有了解，换言之，这些都是开发维护人员非常清楚的细节，我作为普通用户如果只想了解使用区别的情况下，直接自己看代码是非常吃力不讨好的事，因此才会询问有没有知道细节的朋友愿意出来解答下（当然这不是issue，只是discussion，放轻松）

其次，项目的readme我有看过，但是里面介绍分支的描述，大意上是“针对China用户”，“允许采用hack包”这种比较笼统的说法，作为一个普通用户，我的确没法从readme一眼看出，本项目和官方、lean的lede有什么除了package上的区别（为此我查看了package的部分代码，但如果只有这个区别，那本项目应该也不会单独维护一整套代码了，对吧？）。或许readme上可以标注下内核、架构支持、模块方面以及benchmark性能上的区别？

最后，仍然感谢你的回帖，虽然没解答我的问题，但至少让我的帖子不会直接沉掉:)

**Comment by science2468** [comment]
都一样，没区别。要真说区别就是使用人群的需求不同。

**Comment by asiaon123** [comment]
兄弟 这个分支的维护者们怨气都比较大 不能惹

**Reply by Glunefish** [reply]
哪个分支

**Comment by chengnan049** [comment]
> 兄弟 这个分支的维护者们怨气都比较大 不能惹

不至于天灵人还是很好的

**Comment by CrazyBoyFeng** [comment, marked as answer]
作为用户的观感：
lean 的固件有时会打包进闭源的组件。lean 的固件所基于的 openwrt 分支版本一般比较老。
immortalwrt 基本上是开源的。一些流行的闭源驱动或软件都没有被 immortalwrt 收录。这是 immortalwrt 与 lean 的 lede 的主要区别。
与 openwrt 相比，immortalwrt 有时会对系统、软件包进行一些修改。比如合并一些被认为是有优化效果的补丁，在界面上添加 cpu 负载、温度之类的信息，无线开启 256qam（但这个功能的兼容性不怎么好，而且好像没有针对机型优化，比如 mt7603 打开之后就连不上）。软件仓库会收录一些开源但 openwrt 仓库没有的应用。有的还会增加一些对系统优化的选项。还有修改默认参数以优化性能。比如，以前 openwrt 默认不漫游。如果打开漫游，默认的协议是 ft over ds。这个模式不被大多数手机兼容。而在 immortalwrt 中，默认打开漫游，协议默认值被预置为兼容性高的 ft over air。
此外，immortalwrt 通常还会先于 openwrt 提供一些后者尚未提供支持的机器的固件。

默认情况下 immortalwrt 打包的固件与 openwrt 的区别如下：

immortalwrt 额外添加了：
* block-mount 分区挂载。
* kmod-nf-nathelper 分流过滤插件，仅适用于 ipv4。可能被一些 VPN、流量重定向、过滤软件用到。大多数流量重定向和过滤软件使用防火墙规则实现，故该插件可能在大多数情况下用不到。
* kmod-nf-nathelper-extra 前述插件的协议扩展插件，但仍然仅适用于 ipv4。
* luci-app-opkg 软件包管理界面。其实不用写进打包列表。这是 luci 的其中一个依赖，安装 luci 时会自动安装。但是只安装 luci-app-opkg 还是没有中文，所以应该替换为安装 luci-i18n-opkg-zh-cn。
* luci-compat 用以兼容古代 luci 软件包的库。如果不使用很久以前开发的 luci 软件包，那么可能并不需要这个软件包。
* luci-lib-base luci 接口的基础库。其实不用写进打包列表。这是 luci-base 的其中一个依赖。安装 default-settings-chn 时会安装依赖 luci-i18n-base-zh-cn，进而安装依赖 luci-base。
* luci-lib-fs 该软件包没有介绍，我不知道它有什么用。在刷好的机器里，也找不到依赖它的软件。
* luci-lib-ipkg 操作 opkg 的 uci 接口。如果不使用相关脚本，那么这个库是无用的。有些软件，如 luci-app-store，会依赖它，所以安装这类软件时会自动安装这个依赖。因此这个库在绝大多数情况下都是无需额外打包安装的。

immortalwrt 将一些精简版软件替换为完整版：
* dnsmasq 替换为 dnsmasq-full，后者被很多流量重定向和过滤类软件用到。
* libustream-mbedtls 和 wpad-basic-mbedtls 替换为 libustream-openssl 和 wpad-openssl（在低配机器上换为 wpad-basic-openssl），这样做的考虑我估计是因为很多加密代理使用 openssl。但现在很多加密代理使用 go 或 rust 编写，并不依赖系统组件，因此我在我的机器上使用功能和性能都兼顾的 wpad-mbedtls。

immortalwrt 添加了 openwrt 软件库中没有的软件：
* autocore 获取 cpu 状态信息的脚本。
* default-settings-chn 安装依赖 luci-i18n-base-zh-cn，汉化界面；将软件包仓库、ntp 服务器设置为国内服务器。
* ipv6helper 包含 dhcpv6 和 6in4 协议部署脚本。但大多数家庭宽带只用到 dhcpv6，用不到 6in4 隧道，我推荐安装 odhcp6c（dhcpv6客户端）和 odhcpd-ipv6only（dhcpv6服务器）来替代它。

其它我认为打包列表应该修改的地方：
* 删除 opkg 和 luci-app-opkg，替换为 luci-i18n-opkg-zh-cn。安装 luci-i18n-opkg-zh-cn 或 default-settings-chn（依赖关系为 default-settings -> luci -> luci-app-opkg -> opkg）时都会安装 opkg 和 luci-app-opkg。因此并不需要单独列出。
* 删除 luci 和 luci-lib-base。如前所述，安装 default-settings-chn 时会安装依赖项 luci 和 luci-lib-base。
* 将英文的 firewall4 替换为中文的 luci-i18n-firewall-zh-cn。安装时会安装依赖项 luci-app-firewall，进而安装 firewall4。

**Reply by WROIATE** [reply]
非常棒的回答！感谢😆
很抱歉这么久才回复，之前有些忙
我这边先将这个回答标记成answer，给本帖的问题作为参考

**Reply by Icarusradio** [reply]
> ipv6helper 包含 dhcpv6 和 6in4 协议部署脚本。但大多数家庭宽带只用到 dhcpv6，用不到 6in4 隧道，我推荐安装 odhcp6（dhcpv6客户端）和 odhcpd-ipv6only（dhcpv6服务器）来替代它。

捉个虫，客户端应该是 odhcp6c

**Reply by iTofu** [reply]
问一下，[这篇文章所在](https://kzpu.com/archives/5640.html) 是答主的博客吗？

看到这里摘录了答主的内容，但并未标注来源。

**Reply by CrazyBoyFeng** [reply]
to @iTofu:
那不是我的博客。虽然我也希望别人保留原始链接和署名，不过如果影响不是很大的话，我也不希望与人发生纠纷，所以随他去吧。

**Reply by Nehcknarf** [reply]
非常好的资料，按照层主配置调整了一下。
24年6月，luci-lib-base 似乎不再是默认依赖了，建议重新加回打包列表。luci-lib-ipkg应该是iStore的依赖。

**Reply by CrazyBoyFeng** [reply]
> 24年6月，luci-lib-base 似乎不再是默认依赖了，建议重新加回打包列表。

我刚才在 firmware-selector.immortalwrt.org 上试了一下多个机型的 23.05-SNAPSHOT 版本固件，发现默认打包脚本都仍然还包含 luci-lib-base。  
我也观察了 [default-settings-chn 的 makefile](https://github.com/immortalwrt/immortalwrt/blob/75942eaaaba1bca6a52f4060501d5ee81703fd58/package/emortal/default-settings/Makefile#L26)。它的依赖没有改变，仍然还包含 luci-i18n-base-zh-cn，间接依赖 luci-base，然后间接依赖 luci-lib-base。  
所以在默认打包脚本中包含 luci-lib-base 仍然还是冗余重复的。打包 default-settings-chn 时，就已经含进去了。

> luci-lib-ipkg应该是iStore的依赖。

我观察了 [istore 的 luci-app-store 的 makefile](https://github.com/linkease/istore/blob/ef3bbece13098ebca1319460c48dd6a07c08188f/luci/luci-app-store/Makefile#L10)。它确实依赖 luci-lib-ipkg。  
不过，当你自行打包或安装 luci-app-store 的时候，将会自动添加依赖包 luci-lib-ipkg，因此也并不需要额外打包或单独安装 luci-lib-ipkg。  
此外，我之前说 luci-lib-ipkg 是为旧版 luci 提供的接口，这个说法是错的。luci-lib-ipkg 的作用是提供 uci 操作 opkg 的接口。所以类似 istore 这类的程序会用到它。它的功能在 openwrt 18 之前被 luci-lib-base 所包含，而在 openwrt 18 版本中，它的代码被从 luci-lib-base 中所分离成独立的 luci-lib-ipkg 包。  
但这也不改变我之前的结论。我仍然认为 immortalwrt 默认打包 luci-lib-ipkg 是冗余打包。因为极少软件会用到这个包。真的需要这个包时，会自动安装依赖，所以并不需要额外单独打包安装 luci-lib-ipkg。

我主帖中列举了 immortalwrt 中很多这样的冗余打包，都是并不需要单独打包安装，而是安装那些“主体软件”时会自动安装依赖。我不清楚 immortalwrt 维护者额外将它们添加进默认的用意，也许是为了兼容那些漏写依赖的第三方软件包或者没写依赖的脚本？

**Reply by qzydustin** [reply]
总结太棒了！

**Reply by youland** [reply]
> 非常好的资料，按照层主配置调整了一下。 24年6月，luci-lib-base 似乎不再是默认依赖了，建议重新加回打包列表。luci-lib-ipkg应该是iStore的依赖。

被朋友提醒看到这个贴子，那个博客是我的，当时为了传播内容，就转载了 @CrazyBoyFeng 的内容，我现在把链接和来源加上，感谢博主的科普，同时也感谢开发者。对于当时的疏忽向作者道歉！ @iTofu

**Reply by SentimentalEmma** [reply]
ppp 与 ppp-mod-pppoe 是否是依赖关系，前者能否被精简掉呢

**Reply by CrazyBoyFeng** [reply]
ppp-mod-pppoe不依赖ppp，可以精简掉ppp。

**Comment by liyiranlab** [comment]
lede的360t7有nmbm，而且配合immortalwrt的固件的其它路由器漫游效果要比h大的360t7配合immortalwrt的r3g，ax6效果要好，因为我家卫生间的5g信号要差一点，t7使用h大闭源在卫生间经常断网，用lede就要好很多，不知道是不是因为h大的闭源固件只有k，v，没有r设置造成。

**Comment by xuexijiaoben** [comment]
上上楼的兄弟太热心了，赞一个
借贵宝地咨询下，分支master和23.05或21.02之类的分支有什么区别呀，
谢谢啦

**Reply by jh0072110** [reply]
luci版本不同

**Reply by science2468** [reply]
21用iptables 22后用nftables

**Reply by sdttttt** [reply]
这个应该可以参考官方固件的CHANGELOG吧

**Comment by ghost** [comment]
现在的immortalwrt固件防火墙是firewall4，如果添加内容包含mwan3，docker类，商店，多拨，分流，openclash等功能插件时候防火墙一栏会提示混用旧版协议，luci-app-adguardhome如果使用这个功能会提示混用，如果只是编译进固件则不提示。passwall，passwall2，mosdns增加进固件不提示混用，哪怕passwall，passwall2插件本身会自动选择了一些ipatbles功能。此外immortalwrt不兼容argone等常见主题，如果选择了就会打不开设置网页。

**Reply by xlighting2017** [reply]
> 此外immortalwrt不兼容argone等常见主题，如果选择了就会打不开设置网页。

这个结论哪得出来的。。
![telegram-cloud-photo-size-5-6116193773974304824-y](https://github.com/immortalwrt/immortalwrt/assets/25970065/7c91cf4d-c341-40fa-8a65-47fce39f5a87)
![telegram-cloud-photo-size-5-6298845246875876239-y](https://github.com/immortalwrt/immortalwrt/assets/25970065/a145e965-d2dd-4d80-9201-aa945ea3a759)

**Reply by xuexijiaoben** [reply]
实际使用会影响么

**Reply by bbcbbk** [reply]
> 现在的immortalwrt固件防火墙是firewall4，如果添加内容包含mwan3，docker类，商店，多拨，分流，openclash等功能插件时候防火墙一栏会提示混用旧版协议，luci-app-adguardhome如果使用这个功能会提示混用，如果只是编译进固件则不提示。passwall，passwall2，mosdns增加进固件不提示混用，哪怕passwall，passwall2插件本身会自动选择了一些ipatbles功能。此外immortalwrt不兼容argone等常见主题，如果选择了就会打不开设置网页。

luci-app-adguardhome我用了,并不会提示; passwall本来就支持nftables; argon人家github上就写了支持immortalwrt, 我也正在用, 除了没有气泡特效,其他跟lean版的一模一样, 不知道你是怎么得出这些结论的; 

最新的luci-app-mosdns 5.3.3甚至lean lede上用不了, 因为升级到了luci2

**Reply by breeze303** [reply]
提示规则混用不是固件的问题，是某些插件自动创建了ipatbles规则导致的，你可以打开ipatbles规格看看是哪个软件创建的，例如lucky和docker就会默认创建ipatbles规则，只需要关闭和删除ipatbles规则就行，ipatbles -F ipatbles -X，当然还有ipatbles nat的规则也要一并删除，这样就不会混用了

**Comment by Mosney** [comment]
ImmortalWrt是使用体验无限接近主线的一个OpenWrt分支/变体，主要针对墙内用户体验做了很多工作。
在我看来与其它非主线OpenWrt类似物最主要的区别在于ImmortalWrt拥有：
- 有稳定的发行版本
- 提供墙内网络可用的软件包源
- 收录了众多未被包含在主线仓库的软件包
- 提供预编译镜像和自定义镜像

其它的自编译固件与其相比的缺点有：
- 没有包管理源可用，涉及内核依赖的软件包无法安装
- 某些组件停留在老版本，存在潜在风险
- 无法平滑升级

简单来说自编译固件编译好什么样就是什么样，你几乎无法添加或删除功能，也大概率无法升级软件包或保留配置升级到新版本的系统。

**Reply by lccxsusuk** [reply]
aarch64的更新源去那里找老哥

**Comment by nczyw** [comment]
可是为什么immortalwrt 23.05 选上了softether vpn，是无法使用的呢？而lean的选中后编译就能使用，不知道是不是5.15内核的原因，还是防火墙4的原因，反正就是死活用不了，开放端口后，日志显示，500udp，协商加密方式后，等10秒后，日志就显示关闭了，也没有任何错误。

**Reply by science2468** [reply]
去看看日志，问问ai。

**Reply by nczyw** [reply]
服务器 日志显示 通过500端口协商了加密方式，就戛然而止了，10秒后，就显示连接已关闭，没有任何错误提示；客户端提示，在安全层面协商出了问题。

**Reply by science2468** [reply]
那就不知道你怎么设置的了。复查下配置文件把。

**Reply by nczyw** [reply]
查到了了，新 版的，没有 监听ipv4。config中添加就行了

**Reply by ghwwx** [reply]
> 查到了了，新 版的，没有 监听ipv4。config中添加就行了

这个怎么设置，我在最新版的openwrt加了softether vpn一直连不上，是那个listening ：：后面改0.00.0么？感觉无用啊。

**Reply by ghwwx** [reply]
你这个是手动开放端口的吗？我在安装了FW4的openwrt上启动softethervpn直接报错啊。

**Reply by nczyw** [reply]
> 你这个是手动开放端口的吗？我在安装了FW4的openwrt上启动softethervpn直接报错啊。

新版本要手动开放端口，string listening 要加0.0.0.0 ，要监听IPV4，其它的一切照旧
